File: Sources/PylorsAssetMgr/Models/FilterEngine.swift
import Foundation

enum FilterEngine {
    // 匹配 key:value 或 key:"quoted value" 或 bare text
    private static var pattern: Regex<(Substring, Substring?, Substring?, Substring?, Substring?)> {
        /(\w+):(?:(?:"(.+?)")|([^\s]+)|)|(\S+)/
    }

    static let anyKey = "__any__"

    struct FilterCondition: Sendable {
        let key: String
        let value: String
    }

    static func parse(_ input: String) -> [FilterCondition] {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        var result: [FilterCondition] = []
        for match in trimmed.matches(of: pattern) {
            let key = match.1.map(String.init)
            let quoted = match.2.map(String.init)
            let unquoted = match.3.map(String.init)
            let bare = match.4.map(String.init)

            if let k = key {
                let v = quoted ?? unquoted ?? ""
                result.append(FilterCondition(key: k, value: v))
            } else if let b = bare {
                result.append(FilterCondition(key: anyKey, value: b))
            }
        }
        return result
    }

    static func matches(asset: Asset, filters: [FilterCondition]) -> Bool {
        guard !filters.isEmpty else { return true }
        return filters.allSatisfy { matchSingle(asset: asset, filter: $0) }
    }

    private static func matchSingle(asset: Asset, filter: FilterCondition) -> Bool {
        if filter.key == anyKey {
            let query = filter.value.lowercased()
            // 模糊搜索系统标签
            for key in ["name", "suffix", "size_display", "dimensions"] {
                if let tv = asset.getTag(key), tv.lowercased().contains(query) {
                    return true
                }
            }
            // 搜索用户标签值
            for uv in asset.userTags.values {
                if uv.lowercased().contains(query) {
                    return true
                }
            }
            return false
        }

        guard let tagVal = asset.getTag(filter.key) else {
            return filter.value.isEmpty
        }
        return wildcardMatch(text: tagVal, pattern: filter.value)
    }

    private static func wildcardMatch(text: String, pattern: String) -> Bool {
        if pattern.contains("*") || pattern.contains("?") {
            let p = pattern.lowercased()
            let t = text.lowercased()
            // 简单 glob 匹配
            let regexPattern = "^" + NSRegularExpression.escapedPattern(for: p)
                .replacingOccurrences(of: "\\*", with: ".*")
                .replacingOccurrences(of: "\\?", with: ".") + "$"
            return t.range(of: regexPattern, options: .regularExpression) != nil
        }
        return text.lowercased().contains(pattern.lowercased())
    }
}

File: pylors_asset_mgr/core/filter_engine.py
import re
import fnmatch

from pylors_asset_mgr.models.asset import Asset

# 解析 key:value 或 key:"quoted value" 或 自由文本
FILTER_PATTERN = re.compile(
    r'(\w+):(?:(["\'])(.+?)\2|([^\s]+)|)'
    r'|(\S+)'
)


class FilterEngine:
    """解析 key:value 过滤器字符串并匹配资产。"""

    @classmethod
    def parse(cls, filter_str: str) -> list[tuple[str, str]]:
        """解析过滤字符串。

        "importance:high status:\"in review\""
        -> [("importance", "high"), ("status", "in review")]
        """
        if not filter_str or not filter_str.strip():
            return []
        result: list[tuple[str, str]] = []
        for m in FILTER_PATTERN.finditer(filter_str):
            key, _, quoted, unquoted, bare = m.groups()
            if key:
                value = quoted if quoted is not None else unquoted or ""
                result.append((key, value))
            elif bare:
                # 自由文本：所有用户标签值中模糊匹配
                result.append(("__any__", bare))
        return result

    @classmethod
    def matches(cls, asset: Asset, filters: list[tuple[str, str]]) -> bool:
        """判断资产是否匹配所有过滤条件（AND 逻辑）。"""
        if not filters:
            return True
        return all(
            cls._match_single(asset, key, value)
            for key, value in filters
        )

    @classmethod
    def _match_single(cls, asset: Asset, key: str, value: str) -> bool:
        if key == "__any__":
            # 在所有标签值中模糊搜索
            query = value.lower()
            for tag_key in ("name", "suffix", "size_display", "dimensions"):
                tv = asset.get_tag(tag_key)
                if tv and query in tv.lower():
                    return True
            for uv in asset.user_tags.values():
                if query in uv.lower():
                    return True
            return False

        tag_val = asset.get_tag(key)
        if tag_val is None:
            # 仅键匹配 "importance:" → 有该键即可
            return value == ""
        return cls._wildcard_match(tag_val, value)

    @classmethod
    def _wildcard_match(cls, text: str, pattern: str) -> bool:
        """支持 * 通配符的匹配。"""
        if "*" in pattern or "?" in pattern:
            return fnmatch.fnmatch(text.lower(), pattern.lower())
        return pattern.lower() in text.lower()
